### ISSUE TITLE
Not able to get any OAuth or Custom Scopes in Access Token

### ISSUE THREAD

**Issue by hanslai** [aws-amplify/amplify-js#3732]
**Which Category is your question related to?**
Cognito, Oauth2/OIDC  Access Token 

**What AWS Services are you utilizing?** 
Cognito User Pool

**Provide additional details e.g. code snippets**
Using either Auth.signIn or the Vue Authentication Components are not able to get any OAuth or Custom Scopes.  
![oidc+scope](https://user-images.githubusercontent.com/20576491/61875840-194f0700-af1e-11e9-81f8-dad22ff2dd68.png)

Sorry, I only have a image of the source my coworker sent me.
![Amplify+JS](https://user-images.githubusercontent.com/20576491/61876101-a85c1f00-af1e-11e9-9e32-cd3e10b77311.png)

I also tested, I was able to get the OAuth scopes if I use the Token Endpoint in Postman
https://docs.aws.amazon.com/cognito/latest/developerguide/token-endpoint.html

Or do we have to use https://github.com/aws/amazon-cognito-auth-js to get the scopes?  But with Amplify I wonder do we still need to use this amazon-cognito-auth-js library?  Actually, I am confused why there are two JS libraries for cognito.

both #1884 and #1370 have the same problem, which is not solved but closed.  It has been almost a year on this issues already.  Any update on this?

**Comment by haverchuck** [comment]
@hanslai Have you tried using [Auth.federatedSignIn](https://aws-amplify.github.io/docs/js/authentication#oauth-and-hosted-ui) as described here.

**Comment by hanslai** [comment]
@haverchuck, I thought FederatedSignIn is for social provider only, but our App does not use any social providers, we are only use Cognito for now.    Or I didn't understand this correctly?   Is it possible for me to use FederatedSignIn without any social provider but Cognito user pool only?

I understand that Cognito User Pool itself is a valid OAuth2/OIDC Identity Provider, so I was able to get the OAuth2 Scopes from the token endpoint.    Yet,  just not sure how to do it with Amplify.

**Comment by haverchuck** [comment]
@hanslai -  Apologies-  I misread your question.  I'll try to look into this further.

**Comment by hanslai** [comment]
@haverchuck any update on this?

**Comment by haverchuck** [comment]
@hanslai I believe that if you want to use OAuth, you would still call `federatedSignIn()` in order to open the Hosted UI.  The social provider piece would come into play only if you setup third party providers in the User Pool or passed a provider value into the function.  Simply calling `federatedSignIn()` without any parameters would open up Hosted UI with a basic Cognito-backed login screen.    This would enable you to authenticate the user with the userpool without handling their username/password directly in your application, which is one of the primary purposes of OAuth.  Using the signin component, on the other hand, means that you are handling these credentials yourself.

To do this basic OAuth flow, you would need to setup a Hosted UI domain and redirect / signout URLs, which the Amplify CLI can assist you with.

**Comment by Darkxylitol** [comment]
Any Update on this?.

**Comment by vgaltes** [comment]
I´m having the same issue. I can´t get the custom scopes back using Amplify, I only get them back using the custom UI.

**Comment by pmargom** [comment]
Don't understand how this issue is closed when lot of people is requesting this feature!!

**Comment by sandeepsdixit** [comment]
Same problem here... How do I get custom scopes in access token?

**Comment by johngrayuk** [comment]
Given that you can request custom scopes when authenticating using the hosted UI via the Cognito User Pool (as opposed to Facebook), I expect authentication via the Amplify Authenticator React/VueJS component to support exactly the same feature. In both cases (hosted UI or Authenticator component), you are still using Cognito User Pool as an IdP and should be able to request your desired scopes.

**Comment by shukii** [comment]
Any updates on this issue? I'm having the exact same problem

**Comment by DavidWells** [comment]
I'm also unable to protect API gateway endpoints with the `scope` only set to `aws.cognito.signin.user.admin`

For example, (using serverless framework), I'm only able to check for `aws.cognito.signin.user.admin` & not real scopes like `user/add`, `user/delete` etc.

```yml
functions:
  protectedViaScope:
    handler: src/protected.handler
    events:
      - http:
          path: protected
          method: post
          cors: true
          authorizer:
            type: COGNITO_USER_POOLS
            authorizerId:
              Ref: ApiGatewayAuthorizer
            scopes:
              # Hack/workaround: only aws.cognito.signin.user.admin works =( 
              - aws.cognito.signin.user.admin
              # 👇 wont work b/c scopes arent on the accessToken 
              # - users/delete
              # - users/add
```

How are people working around this amplify/cognito limitation? Only using IAM roles?

**Comment by rupertlssmith** [comment]
I think it is also true that not every user wants the admin scope

`aws.cognito.signin.user.admin`

A regular user of an application does not need user admin rights on Cognito, probably just wants some sort of read/write access to parts of an application back-end.

**Comment by tamirrrdorrron** [comment]
This is also an issue for us

**Comment by Lionel1232** [comment]
We are also impacted by this issue.

Very similar use case to @DavidWells. We have protected a few different API Gateway endpoints using different custom scopes and the only scope we ever get back when using AmplifyJS is the `aws.cognito.signin.user.admin` scope.

**Comment by rubene** [comment]
We are also impacted by this issue. Can't write e2e integration tests because the token does not have the necessary scopes to hit api gateway.

**Comment by ddbradshaw** [comment]
+1 on this being an issue for our team. 

Amplify is useless if we want custom login UI and use API Gateway with custom scopes.

**Comment by chrismillah** [comment]
+1 also experiencing this issue

**Comment by hasanbaloch** [comment]
This issue is not just limited to Amplify. Amazon is forcing us to use to the Hosted UI.

No matter which cognito-app-client-id do we sign in with using SDKs, the resulting access token always contain the admin scope (aws.cognito.signin.user.admin). This is not a feature request but a **bug report** (if not a security loophole). What is the significance of cognito-app-client when its assigned scopes are never fetched in the access token?

**Comment by elorzafe** [comment]
Amplify is an open source library that allows customers to interact with Cloud Services such as Amazon Cognito.

Amazon Cognito provides a mechanism to get different scopes on the access token. This mechanism is only supported on OAuth flow using Cognito HostedUI. We’ve communicated the feedback to the Cognito team for their backlog, but I would also encourage you to open a case through your support channels to provide direct feedback outside of the GitHub forums.

**Comment by elorzafe** [comment]
Amplify is an open source library that allows customers to interact with Cloud Services such as Amazon Cognito.

Amazon Cognito provides a mechanism to get different scopes on the access token. This mechanism is only supported on OAuth flow using Cognito HostedUI. We’ve communicated the feedback to the Cognito team for their backlog, but I would also encourage you to open a case through your support channels to provide direct feedback outside of the GitHub forums.

**Comment by DavidWells** [comment]
Any word from Cognito team on this? 

Many folks are experiencing this https://github.com/aws-amplify/aws-sdk-android/issues/684

Thanks for your help

**Comment by AndresMWeber** [comment]
We are using it as expected: to have custom authorization per token via scopes for our service without using the hosted UI (to have a more seamless login experience), however we had to home brew it for now.  It would be amazing to get this functionality so no one has to hand roll anything that shouldn't be missing.

**Comment by tobilg** [comment]
@AndresMWeber could you elaborate how you created your home brewn solution?

**Comment by AndresMWeber** [comment]
> @AndresMWeber could you elaborate how you created your home brewn solution?

It's tailored to our specific use case, however we validate with an API Gateway Lambda authorizer by doing our own Cognito queries against additional headers we send with the request.  So really it's not that we solved the scopes issue itself, just bypassed it.

**Comment by mattmarcum** [comment]
Is someone going to reopen this or should we create a new bug report...?

**Comment by rupertlssmith** [comment]
> Is someone going to reopen this or should we create a new bug report...?

The problem is that the bug is not in amplify, it is in cognito which is not an open source project. I don't know where we can get visbility into what is happening with cognito, if will be fixed, if anyone is even looking at it or cares.

**Comment by hheavener-kyd** [comment]
This is the biggest letdown I've had with AWS. I spent a couple of weeks learning Cognito and building various functions with Amplify to test out how everything works and then come to find out I can't even control the scope of the access token. How can you provide an IDP platform without the ability to define the scopes of the access token? This is complete nonsense and now we're having to investigate other platforms.

@haverchuck are you still a member of the team working on Amplify? Can you give us any updates on this situation? A lot of people here are desperate for an update.

**Comment by jj0b** [comment]
I also just discovered that if I use Amplify without the hosted UI that I'm not able to get scopes with the access token. This makes Amplify an unusable solution for me, as I expect it will for anyone else needing to use scopes to gate access to their resources. 

I can get an access token with scopes from the Cognito authorization server so it seems it should be possible to have Amplify get them as well. Until then, it seems that this limitation in Amplify (and from what I can see it is an issue with Amplify, not Cognito) is just encouraging people to implement a less than ideal solution for authorizing access to their resources.

**Comment by DavidWells** [comment]
> I can get an access token with scopes from the Cognito authorization server

@jj0b interesting... How are you doing this? 😃

Also, are you using the raw cognito sdk's in your client instead of `amplify`'s `Auth` package?

**Comment by jj0b** [comment]
@DavidWells in my client that is using Amplify I'm using the Auth.SignIn method, not raw Cognito SDK, which gets me the token but not the scopes. 

In a separate client that needs to access the same resources I am not using Amplify at all; instead I am using an OAuth 2 Client Credentials Grant flow (this client is a Raspberry Pi) and I am able to get the token as well as the scopes in that case by making a request to the Cognito authorization server.

**Comment by SeFeX** [comment]
Hello @jj0b ! I am using amplify sdk for javascript on the frontend and I have another client for the backend in spring boot, and it is configured as resource server, can you explain the OAuth2 client flow to get user authentication information? The access token that the frontend sends me does not have enough access to obtain the user's information from the url xxxx.amazoncognito.com/oauth2/userInfo

**Comment by jj0b** [comment]
@SeFeX the OAuth 2 Client Credentials Grant flow did not give me access to user authentication information. That flow is for machine-to-machine authorization, which is what I used it for. It returned an access token and scopes that I used to authorize between a backend application running on on a machine, with a service on AWS.

**Comment by SeFeX** [comment]
@jj0b So that token I use to obtain via api or sdk aws the user information of cognito?

**Comment by elorzafe** [comment]
I reopened this is issue, and mark it as a feature-request for the service team.

**Comment by elorzafe** [comment]
Currently from the client (Mobile and Web) the only option available to access custom scopes on the access token is using Cognito Hosted UI.

if you use `Auth.signIn` it will return the default scope.

The scopes that appear on `oauth` param are only valid when using Cognito Hosted UI.

Can anyone mention on this ticket, what is the minimum thing that will allow you to use Cognito Hosted UI?

**Comment by elorzafe** [comment]
@DavidWells 

I want to understand more how the scope from the token will protect your API. Because what I understand is that scopes are configured on the App Client settings and they are not specific to the user unless I am missing something.

Wouldnt be more useful to use AWS IAM to restrict access to the endpoint/path?

**Comment by SeFeX** [comment]
@elorzafe

Hi! I would like to answer in my case, what I understand is that IAM is the user system for the AWS root account, but a user management system is needed for an app developed by itself, with specific accesses, it would be the database of users with their respective authentication, it is more efficient for cognito to maintain authorization and we only validate the token information from the backend, that is what I am trying to do, I see that it could work with a lambda function, but it would be one more layer on aws and not directly in the backend of the application

**Comment by ravenscar** [comment]
@elorzafe I will attempt to explain how I think scopes would help us, and also what hosted UI could do better.

We are currently looking to migrate two user bases from other systems, one is LDAP and the other is from auth0. These represent two different sets of users for us and are owned by different business units. 

The LDAP users use a desktop app, it already has a login screen and we don't want to change how it looks. We would like this app to use the cognito-idp apis and have scopes available based on the client. Beta users could be moved to a new client with additional scopes not available to typical users.

This client uses some embedded web apis which require certain scopes. With other oauth implementations it has been possible to reduce the scope of the access token on a refresh, we would like to do this so the main refresh token can be used to generate a temporary access token with a single scope for a web api on an access server.

We allow access to an API base on a iss/scope whitelist. If you have a permitted iss and the right scope you can access this API, without the ability to use scopes. Without the ability to set scopes in the api calls you have to use hosted ui, and even if you can get past the look and feel of it (which has been a constant constant source of complaints) then it has its own set of limitations.

Allow pre-setting the email/username/phone. Both sets of users can use the same apps on a website. It would be nice ask for their email address first before redirecting to the hosted UI, then just ask for their password on the hosted UI. As it is we have to ask them, direct to the appropriate hosted UI, then they have to fill out their email address again. How frustrating this must be for users.

Silent refresh. Hosted UI must start supporting silent refresh! Storing the refresh token in local storage as amplify does is nuts! How can anyone take cognito seriously as a security product whilst you are advocating this ridiculousness!

Hosted UI error messages should be customisable.

Support passwordless. The users moving from auth0 are sent passwordless links to their mail. These have a seven day expiry, the AWS custom auth flow has a 3 minute expiry so this is not a great fit. It would be nice if Hosted UI would support passwordless links with a longer expiry. Or even passwordless links out of the box.

Fix device tracking. At the moment if you want to use device tracking then you cannot use USER_PASSWORD_AUTH properly as it is incompatible with device tracking and therefore the refresh token cannot be used.

Use SRP. The _ENTIRE_ reason to use SRP is so the password is never transmitted in plain text even over SSL. However I note that when ALLOW_USER_SRP_AUTH is enabled and ALLOW_USER_PASSWORD_AUTH is disabled... The password is still send across the wire to the HOSTED_UI. I would love to know why the cognito team wants people to jump through so many hoops to use SRP (which is mandatory if you are using the cognito-idp api with device tracking turned on in the pool) when they don't even bother to use it themselves, and lazily send the password themselves.

Document things please. The USER_SRP_AUTH method is completely undocumented. It does not follow the RFC as it uses a different value for the generator (02 when the RFC has 05), the hash padding is weird (typically its padded to the size of the prime number, but not in cognito) and is completely incompatible with every compliant javascript implementation of SRP I found outside of the cognito-identity-js one. Go put "Caldera Derived Key" into google and find the top dozen results being people frustrated trying to use USER_SRP_AUTH. 

Anyway, I hope this helps the Amplify and Cognito teams understand a few of the shortcomings of their products from the perspective of someone who has been instructed to use them.

**Comment by hanslai** [comment]
@elorzafe Scope value can protect WebApi because **Cognito's Resource Server's Scope**s are returned as scope in Access Token by Cognito.

API **have to** examine if the Access Token contain this resource in its scope before it allow authorization to the Api resource.

Returning the custom scope is particularly important for Cognito, because Cognito Access Token **does not include "Aud"**, (There are issues complaint about this for over a year, but nothing is done also.), if "Aud" is include, API can just check if Access Token contain the correct value of "Aud", then us as programmer can write a lot less code to check authorization. 

and I agree with @ravenscar that Cognito's different flow are very poor documented, that makes Cognito very difficult to use.

**Comment by ravenscar** [comment]
It's so weird that AWS have implemented the ability to use scopes to protect resources in API Gateway but not set them up to be useful in Cognito. So if you are using a compliant authorization server then you can have a finer grained control over scopes and then all good with APIGW but if you're using User Pools then:

* you can't choose scopes on login
* you can't use refresh to downscope as AWS didn't bother to implement that
* you can use any scopes when not using Hosted UI.

It's like the Cognito team took all the practices people were using for years and threw them out the window.

* Have an audience on the token? Why bother!
* Use scopes? Here's a 50% implementation that works for no one.
* Implicit flow for webapps? No no no, use the code flow so you can get a refresh token.
* Silent refresh? Nah, just store the refresh token in the browser, it'll be fine.
* Pass the access token to the API? Madness! the client side only id token should be enough for anyone.
* Use SRP? Sure, let's also sent the user's password.

I especially like the little quirks like when you turn on mandatory device tracking that Cognito won't let you use a refresh token without sending the device id... But it gives you access tokens without you confirm a new device.

**Comment by hanslai** [comment]
Even I am still trying to use Cognito myself because I try to use the same cloud provider for everything, but actually Azure AD B2C has the same price now.  50,000 user free per month.   And I learned people have switched to other IDP due to the fact that Cognito is difficult to use especially for Blazor programmers.

**Comment by tobilg** [comment]
Basically @ravenscar summed up all the concerns an problems we're having. Thanks for the write-up!

**Comment by rolandnyamo** [comment]
ugh. been chasing an issue for hours, almost going crazy. but this was it. I wonder what we can do to prioritize it. It's not like cognito is opensource. In the meantime, I guess add the user.admin scope to all API calls?

**Comment by michaelbrewer** [comment]
@ravenscar sums it up well. I really hope the Congnito team has the time to start looking at this growing backlog of issues. And it is not like we have much of a choice.

**Comment by tobilg** [comment]
We're currently using a kind of workaround where we use id tokens instead of access tokens, because they can be updated/verified/etc. with the PreToken Lambda trigger. If I understand it correctly, if a user knows which scopes are available within the hosted UI, he/she could assign the scopes to him/herself by just adding them to the /token or /authorize URL, and there's no current way to verify if a user is allowed to have these scopes or not.

This is less than ideal, because we need to use REST instead of HTTP APIs on API Gateway, because the latter doesn't support private endpoints (yet). 

When using REST, you loose the ability to just directly use the Cognito Authorizer, because the builtin one does check for the token type, and will only verify access tokens, but no id token (why this is the case, no one knows)... You can adapt the Lambda integration so that you can add the desired scopes as properties, which then works again with id tokens. But that comes to the prize that there's a lot more IAC to write, and handling the Lambda responses/errors gets much more inconvenient.

So we ended up writing a custom Lambda authorizer, which basically does what the Cognito authorizer does, to just be able to use custom verified scopes in our id tokens.

Is there something we misunderstood here? This is much too complex/incomplete to be trustfully run in such an important area like authorization/authentication.

**Comment by disbelief** [comment]
I wasted 3 days trying to figure out why the access tokens that Amplify Auth fetches couldn't access my API Gateway methods. Naturally, I assumed it was a problem with my infrastructure config (which is non-trivial), but it seems I should have just searched the Amplify bugs list first.

And yes this seems a lot more like a **bug** than a **"feature request"**.

Using Amplify + Cognito to authorize API Gateway methods is one of the most common and AWS-promoted scenarios. I'm utterly baffled as to why this doesn't already work, and has not worked for years.

I'm also baffled this limitation isn't mentioned in any of the docs. In fact there are places in the docs that indicate it works as expected (eg. ["If the client doesn't request any scopes, the authentication server uses all custom scopes associated with the client"](https://docs.aws.amazon.com/cognito/latest/developerguide/token-endpoint.html)).

`</rant>`

@hasanbaloch and/or @rupertlssmith you mentioned above that the bug lies in Cognito. Just to confirm, this means that the _only_ way to get an access token with other scopes is via the Hosted UI? Ie. if I try to implement the [Cognito token API request](https://docs.aws.amazon.com/cognito/latest/developerguide/token-endpoint.html) myself with an HTTP library I'd have the same problem?

@tobilg thanks for describing your workaround. I'm curious why you felt that the `preToken` Cognito Lambda trigger is more complex/error-prone than a custom Lambda Authorizer? If you hard-code the scopes in the `preToken` Lambda trigger at least you can avoid the user guessing them and the Lambda just blinding adding them. What are the failure modes here?

**Comment by ravenscar** [comment]
@disbelief I may be able to answer these questions as I have been writing a library for dealing with cognito and I have a few integration tests which tries stuff like this.

If you get a refresh token via the API calls, rather than Hosted UI, you will not be able to use it in the `https://dewfewfewfewf.auth.ap-southeast-2.amazoncognito.com/oauth2/token` endpoint. You will get a response such as `{"error":"invalid_client"}`. To use that endpoint you must obtain a refresh token via the hosted UI's code_flow.

As far as I am aware there is no way to get any OAuth scopes other than via the Hosted UI. I and my team have tried many many things and also contacted AWS support using our enterprise plan, and we have been told the same by them.

Also please note that tobilg's workaround only modifies the id token, not the access token, so it is useless for OAuth2/OIDC. Scopes have no purpose on the id token, e.g. you could call it anything at that point.

For me I can do without scopes but not being able to use something like code_flow with the API is a real pain when you are doing mobile development with your own embedded UI, as you have no secure way to deliver the tokens back to your app.

As for what are the failure modes? Well who knows, all this is pretty far of the reservation. Cognito's lack of flexibility has caused some people to go to [extreme lengths](https://stackoverflow.com/questions/49012091/aws-cognito-authorization-code-grant-flow-without-using-the-hosted-ui) to get basic things working.

If this service was provided by another company and not amazon I am fairly sure they would have almost no customers given the offering they provide.

**Comment by disbelief** [comment]
Thanks @ravenscar for the run down, saved me surely quite a few hours (or days) trying to implement myself. I did try a couple more things without any luck.

First I experimented with hitting the token endpoint directly in Postman but couldn't get it working (`invalid_client` was a common response).

I also tried the preToken trigger that tobilg mentioned, but as you say the scopes only get added to the id token, and API Gateway doesn't accept that token regardless.

In the end I've decided to "roll my own" custom authorizer lambda, which talks to Cognito directly to verify ID tokens. I don't have fine-grained scopes in my system — all users can access the same APIs — so it is enough for me to just verify the user is authenticated.

It goes against security best practices and defeats the purpose of OAuth2 + OIDC, but it seems AWS people are comfortable with offering their customers this level of compliance.

**Comment by tobilg** [comment]
@disbelief / @ravenscar It's not entirely correct that API Gateway can't be used to verify scopes in id tokens. It works as I mentioned, but to the cost that you have to change from Lambda proxy to Lambda integration method and the Lambda function responses as well, and write more IAC.

See https://medium.com/@lakshmanLD/lambda-proxy-vs-lambda-integration-in-aws-api-gateway-3a9397af0e6d for an idea.

Example in Serverless (would validate if the scope `test/test` is in the id token's claims):
```yaml
resources:
  Resources:
    ApiGatewayAuthorizer:
      Type: AWS::ApiGateway::Authorizer
      Properties: 
        Name: cognitoAuthorizer
        Type: COGNITO_USER_POOLS
        IdentitySource: method.request.header.Authorization
        RestApiId: 
          Ref: ApiGatewayRestApi
        ProviderARNs: 
          - '#{CognitoUserPool.Arn}'
 
functions:
   testFunction:
    handler: src/test.handler
    memorySize: 128
    timeout: 3
    events:
      - http:
          path: test
          method: get
          cors: true
          idTokenAuthorizer: true
          integration: lambda
          request:
            template:
              application/json: '{ "test/test": "$context.authorizer.claims[''test/test'']" }'
          response:
            headers:
              Access-Control-Allow-Origin: "'*'"
              Access-Control-Allow-Credentials: "'true'"
              Access-Control-Allow-Methods: "'DELETE, POST, GET, OPTIONS, PATCH'"
              Access-Control-Allow-Headers: "'Content-Type, X-Amz-Date, Authorization, X-Api-Key, X-Amz-Security-Token, X-Amz-User-Agent'"
          authorizer:
            authorizerId: '#{ApiGatewayAuthorizer}
```

I agree that it's not ideal to use the id token, but at least for me that's a workaround that works (in conjunction with a PreToken trigger and a custom authorizer where I define on route level the necessary scopes to access the respective route), I'm not planning to integrate into other systems via OAuth2/OIDC anyway.

Having said that, it's a pity that Cognito/API Gateway can't currently do this out of the box, and I personally know people who have switched to Auth0 or the likes because of that.

**Proposed solution:**
Make the PreToken trigger usable for access tokens as well, or create a new trigger with which scopes can be added/verified in access token.

I also proposed this long ago via enterprise support channels, Twitter discussions with AWS employees and also via the AWS Community Builders, to no avail unfortunately.

**Comment by disbelief** [comment]
@tobilg that actually doesn't sound too bad, thanks for sharing your solution. I'm going to give this a shot as it seems like less custom authentication/authorization code to write vs. rolling my own lambda function to verify, authenticate, and authorize JWT tokens.

**Comment by ravenscar** [comment]
@disbelief if you don't have fine-grained scopes, and are just using cognito to authenticate (this is what we are doing) have you just considered using API Gateways built in access token authenticator and telling it to use the `aws.cognito.signin.user.admin` scope? This way you probably don't need to do anything fancy.

**Comment by disbelief** [comment]
@ravenscar that's a good point. I probably should have just done that. But I've learned ever so much about velocity templates now — silver linings 😂

**Comment by xyklex** [comment]
I've been around this specific issue several times, and I don't know if this could help but I'll still share it in case it is useful to introduce into the SDK or a new client, maybe useful to @ravenscar @disbelief

So because of the lack of allowing setting custom scopes in Amplify SDK, I did try to make use of the OAuth endpoints to see if I can reproduce the Authentication Code Flow from the shell and not directly from the browser to understand the process. 

These are the steps:
1. Call the /oauth2/authorize endpoint
```
curl -X GET "https://bw-customers-login.auth.us-east-1.amazoncognito.com/oauth2/authorize?response_type=code&
client_id=72vvq39ev9op0cqn102jp2a51f&
redirect_uri=http://localhost:8081/&
state=429EA3CA-B8F5-45F9-B423-6E85DF7E2B26&
scope=openid+profile+email" -c cookies.txt
```
which gives you a redirect answer (302) but I store the cookies return with `-c cookies.txt`
```
#HttpOnly_bw-customers-login.auth.us-east-1.amazoncognito.com   FALSE   /       TRUE    1582913132      csrf-state-legacy       ""
#HttpOnly_bw-customers-login.auth.us-east-1.amazoncognito.com   FALSE   /       TRUE    1582913132      csrf-state      ""
#HttpOnly_bw-customers-login.auth.us-east-1.amazoncognito.com   FALSE   /       TRUE    0       XSRF-TOKEN      7e6f047a-8080-4c8c-a990-d9b85b0b0389
```

2. Then I do a POST request to the value in `Location:` header
```
curl -i -X POST -H 'Content-Type: application/x-www-form-urlencoded' "https://bw-customers-login.auth.us-east-1.amazoncognito.com/login?response_type=code&
client_id=72vvq39ev9op0cqn102jp2a51f&
redirect_uri=http://localhost:8081/&
state=429EA3CA-B8F5-45F9-B423-6E85DF7E2B26&
scope=openid+profile+email" -d "username=<email>&password=<password>&_csrf=<XSRF-TOKEN>" -b cookies.txt -c cookies.txt
```
replace:
`<email>` with the registered email
`<password>`  with the corresponding password
and `<XSRF-TOKEN>` with the corresponding value from the cookie file
The `state=` value does not matter too much here and could be anything

This will return you a new Location URL in the response headers with the code  value

3. Exchange code for access tokens
```
curl -X POST -H 'Content-Type: application/x-www-form-urlencoded' "https://bw-customers-login.auth.us-east-1.amazoncognito.com/oauth2/token" -d 'grant_type=authorization_code&client_id=72vvq39ev9op0cqn102jp2a51f&redirect_uri=http://localhost:8081/&code=97b42149-13db-4dcd-8f11-9b108cc94b0a'
```

This will return you an `id_token`, `access_token`, and `refresh_token`

**Comment by tobilg** [comment]
@xyklex IMHO the Cognito Auth API is documented at https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-userpools-server-contract-reference.html so I'd say this is known information.

**Comment by zoellner** [comment]
Another problem adding to this is that the hosted UI doesn't properly support two factor authentication that was finally added recently. So there is no solution that actually works to use custom scopes and 2FA because one requires you to use the hosted UI and the other requires you to handle the login yourself.

**Comment by theWebKeyGuy** [comment]
Just want to share this with everyone that may need a better understanding. This was shared with me by the Amplify Staff on their discord. This helped me get a better understanding of how things were meant to be and why.

This is how it works behind the scenes: 
https://www.youtube.com/watch?v=1N0lNLHYGVs

More Info:
https://docs.amplify.aws/lib/auth/overview/q/platform/js/
https://docs.amplify.aws/cli/auth/groups/

**Comment by ahfx** [comment]
For anyone having this issue with Postman and the OAuth 2.0 Authorization type, you also need to add the required scope in the fields provided (I spent my fair share of time googling before realizing this):

<img width="558" alt="image" src="https://user-images.githubusercontent.com/13871698/152697682-753d973c-5894-4673-824d-ea47c3d5614a.png">

**Comment by estebane-frb** [comment]
Hi there, any updates on this? or estimates on when a fix is planned? I want to use custom scopes with the Amplify UI, not the Cognito Hosted UI.

**Comment by Tran-Minh23** [comment]
It's funny when you seeing someone faced this issue in 2019, and in 2022 you still got the same problem without any solution...

**Comment by tobilg** [comment]
I guess that's called "Customer Obsession"...

[AWS Leadership Principles](https://www.amazon.jobs/en/principles)

**Comment by jcardus** [comment]
Suffering with this also... Using cognito / amplify was one of the worst decisions of my life...

**Comment by viniarruda** [comment]
Hello guys, any news about this issue?

**Comment by arthurgurov** [comment]
One day you decide to become a certified AWS expert... you spend a lot of time and money to achieve that. Once certified, you share your joy with all your friends and clients. You dream about modern cloud solutions with super cool architectures. You think you know everything needed to build a cool solution, not only because of fancy marketing words - you think you know this platform.

And then... then you see such sh## like this... when a super obvious thing is not working and no one plans to fix it (only 3 years passed from the initial post). I was really disappointed when I noticed almost zero customization of the hosted UI. But this... this is just ridiculous.

Now I'm at a point where I regret the decision to use Cognito because it's a real blocker for us. I have to step back and re-architect the system to use a more bullet-proof solution than Cognito.

**Comment by stewberticus** [comment]
Are there any valid work arounds? I need a custom scope in my access token

**Comment by Tran-Minh23** [comment]
> Are there any valid work arounds? I need a custom scope in my access token

I think there is a way, you can try add your users to a specific group in Cognito. The access token can return the user group, and in the backend you can filter that claim. I hope it fits your use case

**Comment by SonofNun15** [comment]
As mentioned with @ravenscar, the major issue for us with the Hosted UI is that we need our users to enter their email into our portal before we redirect them to the login page so that we can migrate existing users to our Cognito user pool. Requiring them to enter their email twice is absurd and there appears to be no way to inject this into the Hosted UI. Just no customizability there. But then this issue with custom scopes is just baffling.

**Comment by matthew-mcdermott** [comment]
I see people saying you can use the hosted ui, but for me that doesn't even work. No matter what I try, I get no scopes. Any insight from anyone? 

Also, I hope the AWS team can handle this soon :/

**Comment by lisadesouza89** [comment]
I agree, I'm using the Amplify frontend for the simple integration - now I was looking to plug into other oauth based systems and my access token is missing the openid scope. Help @aws-amplify-ops

**Comment by abdallahshaban557** [comment]
Hello everyone, we are discussing this request with the Cognito team. We realize this is important for you, and we will keep this issue updated when we have progress!

**Comment by lisadesouza89** [comment]
> Hello everyone, we are discussing this request with the Cognito team. We realize this is important for you, and we will keep this issue updated when we have progress!

Any update on this? @aws-amplify-ops ?

**Comment by AndroidDoctorr** [comment]
I would also really like an update on this.

**Comment by abdallahshaban557** [comment]
The Cognito team is investigating supporting this feature. We do not have exact timelines yet - but we will share them when we have more details!

**Comment by cracanut1** [comment]
Also waiting for this to be fixed.

**Comment by whimzyLive** [comment]
![giphy (3)](https://user-images.githubusercontent.com/22066605/232113816-64c0f979-2e88-4e66-8788-366ca5cbe6e0.gif)

**Comment by ym-herby** [comment]
I am also blocked on this issue.  In my use case when logging in using Google as the Identity provider I get a token with the openid scope included, which makes sense. If I user username/password login which presumably uses Cognito itself as the identity provider then the only scope that is included is the `aws.cognito.signin.user.admin`.  How do we configure Cognito to behave like any other OIDC provider and issue an access token with the **openid** scope?

**Comment by abdallahshaban557** [comment]
Hello everyone, this issue is currently a limitation with Cognito not supporting the retrieval of custom scopes with SRP login (logging in with Email/Username and password). We are working closely with their team to get it resolved, and we will provide an update when this feature is available. We realize that this is an essential feature for providing granular Authorization, and it is a top priority for us to expose that functionality once Cognito makes it possible. We will provide updates here when we have clarity on our timelines.

**Comment by adambeer** [comment]
@abdallahshaban557 Sorry that you have to take the brunt of this but "top priority" shouldn't take another month. People are paying money for Cognito and Amplify. This is embarrassing.

**Comment by panterozo** [comment]
Same here.. Is it possible to have any update on this?

**Comment by abdallahshaban557** [comment]
Hello @adambeer @panterozo  - we do not have an update yet. We will provide an update when we have the next steps clear with Cognito.

**Comment by PeelyZenobe** [comment]
> Hello everyone, this issue is currently a limitation with Cognito not supporting the retrieval of custom scopes with SRP login (logging in with Email/Username and password). We are working closely with their team to get it resolved, and we will provide an update when this feature is available. We realize that this is an essential feature for providing granular Authorization, and it is a top priority for us to expose that functionality once Cognito makes it possible. We will provide updates here when we have clarity on our timelines.

Thank you @abdallahshaban557  for letting us know this is a limitation with Cognito, and not Amplify. This is helpful when it comes to ultimately deciding if we want to use Amplify in our next project, and trying to find a workaround.

Can I ask, is the limitation you mentioned anything to do with the large callout at the top of this page? https://docs.aws.amazon.com/cognito/latest/developerguide/user-pools-API-operations.html For ease, i'll paste it here
>  Note
Access tokens from user pools API authentication only contain the aws.cognito.signin.user.admin scope. If you instead want to generate an access token with additional scopes, for example to authorize a request to a third-party API, **authenticate your users through your user pool endpoints**.

Emphasis mine, to clear up any confusion it means to use the AWS Hosted UI to handle Login, instead of using the User Pools API (Which is what I _assume_ the Auth.SignUp and Auth.SignIn functions use under the hood). So I assume that is what you are talking about?

**Comment by abdallahshaban557** [comment]
@PeelyZenobe  - yes, you are correct! 

@marcjulianfleck  - they are working on enabling it ASAP. We apologize this has taken so long.

**Comment by PeelyZenobe** [comment]
> @PeelyZenobe - yes, you are correct!
> 
> @marcjulianfleck - they are working on enabling it ASAP. We apologize this has taken so long.

Thank you for getting back to me so fast @abdallahshaban557, I really appreciate it. It's good to hear that the Cognito team are working on enabling it ASAP :) To free you up from being a constant go-between, is there an Cognito representative in this Issue thread? Or another Issue we can track the Cognito teams progress on? As you said, it's not an Amplify problem.

**Comment by jamesh38** [comment]
@abdallahshaban557 Any luck this month?

**Comment by shaktiks** [comment]
any update on this? It seems like aws is going down the drain, time to move to GCP.

**Comment by cwomack** [comment]
Hey, @jamesh38 and @shaktiks 👋.  We can't announce an exact release date just yet, but this feature is actively being worked on and planned for an upcoming release.  As soon as we can communicate an actual day it will be available, we'll be sure to do so!

**Comment by jamesh38** [comment]
Hey that’s awesome to hear. Thanks a lot Chris.

**Comment by Ataraxia009** [comment]
hey @cwomack thanks for looking into this
Just want to confirm ->
the new new release will also allow custom scopes to be sent in the access token for CUSTOM_AUTH flows right?
Specifically I am using the lambda trigger auth challenges and the [defineAuthChallenge](https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-lambda-define-auth-challenge.html#cognito-user-pools-lambda-trigger-syntax-define-auth-challenge) lambda trigger. 
The response is quite limited in what to feed the access token.

**Comment by danfhernandez** [comment]
Any update on this? Something I'm also running into unfortunately. I'm very close to choosing to migrate off of amplify and roll auth myself.

**Comment by PeelyZenobe** [comment]
> Any update on this? Something I'm also running into unfortunately. I'm very close to choosing to migrate off of amplify and roll auth myself.

Just to be clar, you don't need to "roll auth yourself". You can still just use Cognito without Amplify. Yes you can still only get the custom scopes if you use the Amazon Cognito provided Hosted UI (Which provides all of the Authentication journeys (Signup, Login, Password Reset and MFA) out of the box), but in my investigations it turns out that custom scopes aren't really needed that often, [it's an anti-pattern to use them for User Permissions](https://auth0.com/blog/on-the-nature-of-oauth2-scopes/) and you should be using another system to handle that. Machine-to-Machine stuff can use Cognito just fine to get customer scopes.

**Comment by rupertlssmith** [comment]
Seems to have been fixed on Cognito User Pools: https://aws.amazon.com/about-aws/whats-new/2023/12/amazon-cognito-user-pools-customize-access-tokens/

**Comment by lobaak** [comment]
> Seems to have been fixed on Cognito User Pools: https://aws.amazon.com/about-aws/whats-new/2023/12/amazon-cognito-user-pools-customize-access-tokens/

I'm just confirming that it is now possible to add custom scopes to the access token generated using Amplify. We were able to get this working using the steps found on:

https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-lambda-pre-token-generation.html#user-pool-lambda-pre-token-generation-accesstoken

The lambda function we used is taken from the example on that page:

```node
export const handler = function(event, context) {
  event.response = {
    "claimsAndScopeOverrideDetails": {
      "accessTokenGeneration": {
        "scopesToAdd": [
          "openid",
        ],
        "scopesToSuppress": [
          "aws.cognito.signin.user.admin"
        ]
      },
    }
  };
  // Return to Amazon Cognito
  context.done(null, event);
};

```


Which generates the following access token:


```json
{
  "sub": "xxxx",
  "iss": "https://xxxx.amazonaws.com/ap-southeast-2",
  "client_id": "xxxx",
  "origin_jti": "xxxx",
  "event_id": "xxxx",
  "token_use": "access",
  "scope": "openid",
  "auth_time": 1705450802,
  "exp": 1705451102,
  "iat": 1705450802,
  "jti": "xxxx",
  "username": "xxxx"
}
```

**Comment by sgasior** [comment]
> I'm just confirming that it is now possible to add custom scopes to the access token generated using Amplify. We were able to get this working using the steps found on:
> 
> https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-lambda-pre-token-generation.html#user-pool-lambda-pre-token-generation-accesstoken
> 
> The lambda function we used is taken from the example on that page:
> 
> ```js
> export const handler = function(event, context) {
>   event.response = {
>     "claimsAndScopeOverrideDetails": {
>       "accessTokenGeneration": {
>         "scopesToAdd": [
>           "openid",
>         ],
>         "scopesToSuppress": [
>           "aws.cognito.signin.user.admin"
>         ]
>       },
>     }
>   };
>   // Return to Amazon Cognito
>   context.done(null, event);
> };
> ```
> 
> Which generates the following access token:
> 
> ```json
> {
>   "sub": "xxxx",
>   "iss": "https://xxxx.amazonaws.com/xxxx",
>   "client_id": "xxxx",
>   "origin_jti": "xxxx",
>   "event_id": "xxxx",
>   "token_use": "access",
>   "scope": "openid",
>   "auth_time": 1705450802,
>   "exp": 1705451102,
>   "iat": 1705450802,
>   "jti": "xxxx",
>   "username": "xxxx"
> }
> ```

It requires to use `Advanced security features` and this is very expensive feature as I know...(0.05$).

Is it save to grant `aws.cognito.signin.user.admin` to the user? What use is able to do with this scope?

**Comment by DavidWells** [comment]
I was quite pleased when I saw the announcement that is was possible to modify the `access` token via a lambda function. https://aws.amazon.com/about-aws/whats-new/2023/12/amazon-cognito-user-pools-customize-access-tokens/

But then like @sgasior I realized it requires opting into `Advanced security features`. For access token customization you must enable "AdvancedSecurityMode" which blows up the cost of MAU to 0.05 per user from 0.0055 per user. 

**This is a 809.09% price increase per user. ** 😭

Surely there is a way to enable the new `Pre token generation Lambda trigger` to the `V2_0` signature without this incredible increase in cost. 

We just need this one simple feature, not everything that is included in the `Advanced security features` offering.

<img width="1235" alt="image" src="https://github.com/aws-amplify/amplify-js/assets/532272/fc28d66c-68f0-4181-9f56-e6721ba72e9e">

**Comment by johnpwrs** [comment]
My annoying work around at the moment is to look at the scope, if it equals `aws.cognito.signin.user.admin`, i fetch user info using AWS SDK https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/client/cognito-identity-provider/command/GetUserCommand/

Otherwise, I call the oauth2 userInfo endpoint.

**Comment by johnpwrs** [comment]
My annoying work around is to use AWS SDK  [GetUserCommand](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/client/cognito-identity-provider/command/GetUserCommand/) when the user only has the `aws.cognito.signin.user.admin` scope, and to use the oauth userinfo endpoint otherwise.

**Comment by apekkar** [comment]
Hello! Any status update on this? Is there any workaround?

We just made big refactoring in backend to support custom scopes and after that we realised that our apps made with amplify wouldn't work with it anymore. This is frustrating situation.

**Comment by Meags27** [comment]
> We gave up on Amplify/Cognito. We've tried on multiple occasions to find a way to make it viable with different projects and, ultimately, this is the main issue that keeps us from using it. This thread is 5 years old and still active yet I haven't seen any indication that the team has any plans of fixing this.
> 
> Beyond that, they should not advertise this product as a feature-complete authentication service because it isn't. More and more teams are committing to using this product only to realize it doesn't do as advertised after the build out their applications to it. What a let down.

They fixed it December last year https://aws.amazon.com/about-aws/whats-new/2023/12/amazon-cognito-user-pools-customize-access-tokens/

**Comment by hheavener-kyd** [comment]
@Meags27 thanks for the link and clarification. Unfortunately, as @DavidWells mentioned in his comment, you must enable advanced security features in order to use this which drastically increases the cost and contains a lot of features we do not need. This decision means that this product will continue to not be viable for our team.

**Comment by ChristianSMcode** [comment]
Is there any update on this? the solution as mentioned by multiple users involves a crazy diference in cost per user which I find unaceptable and unviable, however I apreciate that at least there was something made about this issue however is far form the ideal solution

**Comment by cs17** [comment]
Enabling advanced security features for this simple feature sounds crazy. I would appreciate it if the Amplify team could look at this request.

**Comment by alexkates** [comment]
We're in the same boat. After reading this, we are landing where many of you have already. The markup per MAU is way too expensive to justify access token customization.

What's funny is I don't even want customization. I want the scopes that are setup on the user pool app client to simply be on the access token. That's not customization, that feels more like "the way it should work".

For now we are choosing to not use authorization scopes until we can move away from the Amplify JS libraries.

**Comment by 5t33** [comment]
Wanted to link this here because I _think_ it's related: https://github.com/aws-amplify/aws-sdk-android/issues/684 

It seems like at least there's a "fix": https://github.com/aws-amplify/aws-sdk-android/issues/684#issuecomment-1862299552

although it bumps up pricing quite a lot: https://github.com/aws-amplify/aws-sdk-android/issues/684#issuecomment-1937892170

Not sure it's worth the price increase, but there might at least be a way forward.

**Comment by arthurgurov** [comment]
> One day you decide to become a certified AWS expert... you spend a lot of time and money to achieve that. Once certified, you share your joy with all your friends and clients. You dream about modern cloud solutions with super cool architectures. You think you know everything needed to build a cool solution, not only because of fancy marketing words - you think you know this platform.
> 
> And then... then you see such sh## like this... when a super obvious thing is not working and no one plans to fix it (only 3 years passed from the initial post). I was really disappointed when I noticed almost zero customization of the hosted UI. But this... this is just ridiculous.
> 
> Now I'm at a point where I regret the decision to use Cognito because it's a real blocker for us. I have to step back and re-architect the system to use a more bullet-proof solution than Cognito.

So... 2 years have passed, and we still need an adequate solution (the new bill multiplier feature is not an option). The good thing is that I have this thread pinned, and now I can make the right decision not to use Cognito at the beginning of the new project :D

**Comment by adrian1chia** [comment]
is really bullshits coming with AWS. This is just an example if a simple thing that never can be done right. I am also stucked now thinking what to do

**Comment by kaleb-itm** [comment]
First i had issues with Cognito requiring saml 2.0 redirect binding and the idp i need to add only supports POST
Now the access token does not have the required info to be able to easily get the users email address (how we id users)
I gave up and im now just using the Id Token.

Next time im just going to use keycloak

**Comment by josephharkins** [comment]
It looks like they have gotten their act together and sorted this, new better pricing tiers, access tokens incl.

https://aws.amazon.com/cognito/pricing/

**Comment by AllanZhengYP** [comment]
Hi everyone,

Apologize for delayed response. We want to share an update from the service side: Cognito service now provides [pre-token generation trigger](https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-lambda-pre-token-generation.html#user-pool-lambda-pre-token-generation-accesstoken), and they suggest to use the trigger to add additional scopes.

I will put this issue to pending closing for now. Please comment here if you think this feature is not sufficient for your use case. Thank your everyone for your inputs!

**Comment by JulianMcOmie** [comment]
Only took 3 years

**Comment by falcolnic** [comment]
5 years in total, damn....

**Comment by ashika112** [comment]
Hi everyone. We have left the [last comment](https://github.com/aws-amplify/amplify-js/issues/3732#issuecomment-2819323661) out for a bit and dont see any comments on this not fulfilling the use case. Greatly appreciate the patience. This feature needed to be supported  Cognito service team. Amplify have little control over this feature and needs no code change for customer to be able to use this feature. 

I am closing out this issue now that Cognito has supported this. If there is any further clarifications or issue while working through this. Please open a new issue ticket with the team. Thanks.

**Comment by theWebKeyGuy** [comment]
> Hi everyone. We have left the [last comment](https://github.com/aws-amplify/amplify-js/issues/3732#issuecomment-2819323661) out for a bit and dont see any comments on this not fulfilling the use case. Greatly appreciate the patience. This feature needed to be supported Cognito service team. Amplify have little control over this feature and needs no code change for customer to be able to use this feature.
> 
> I am closing out this issue now that Cognito has supported this. If there is any further clarifications or issue while working through this. Please open a new issue ticket with the team. Thanks.

I really don't feal there should be an extra cost to this, it's like saying you should charge more for IAM.
